### PR TITLE
Fix line skipping issue in receive_lines method

### PR DESCRIPTION
**Pull request by yugeeklab**
**Which issue(s) this PR fixes:**
Fixes https://github.com/fluent/fluentd/issues/4494
recoverd from #4491 

**What this PR does / why we need it:**
Before this patch, long lines could cause breakdowns in fluentd, potentially posing a vulnerability. With this patch, max_line_size will be integrated into the FIFO, enabling the system to skip lines exceeding the maximum size before executing receive_lines.

**Docs Changes:**

**Release Note:**

### CODE REVIEW

**Comment by daipom** [comment]
@yugeeklab I have pushed the commits for https://github.com/fluent/fluentd/pull/4491#issuecomment-2162508391.

**Review by daipom — APPROVED** [review]
Thanks! LGTM.
I will wait a bit to merge for other maintainers.

**Review by ashie — CHANGES_REQUESTED** [review]
Basically it look good to me, thanks for your contribution!
I've added some nitpick comments about logging.

In addition, it would be happy if we can simplify `FIFO#read_lines` to make easy to read although I have no idea about it yet.
What do you think about these my comments? @daipom

File: lib/fluent/plugin/in_tail.rb [hunk 1]
```diff
@@ -1059,11 +1057,38 @@ def read_lines(lines)
...
             @buffer = @buffer.slice(idx + 1, @buffer.size)
             idx = @buffer.index(@eol)
+
+            is_long_line = @max_line_size && (
+              @skip_current_line || rbuf.bytesize > @max_line_size
+            )
+
+            if is_long_line
+              @log.warn "received line length is longer than #{@max_line_size}"
+              @log.debug("skipped line: ") { convert(rbuf).chomp }
```
[the remark this answers is not in the document]
Comment (daipom) [reply lib/fluent/plugin/in_tail.rb]: Thanks!
Fixed: 955cee5681f2c1cb375364a421ac1bdc2364e811

File: lib/fluent/plugin/in_tail.rb [hunk 2]
```diff
@@ -1059,11 +1057,38 @@ def read_lines(lines)
...
+
             lines << convert(rbuf)
           end
+
+          is_long_current_line = @max_line_size && (
+            @skip_current_line || @buffer.bytesize > @max_line_size
+          )
+
+          if is_long_current_line
+            @skip_current_line = true
```
[the remark this answers is not in the document]
Comment (daipom) [reply lib/fluent/plugin/in_tail.rb:1091]: Thanks!
Fixed: 955cee5681f2c1cb375364a421ac1bdc2364e811

**Review by ashie — APPROVED** [review]
LGTM
Thanks!

**Comment by daipom** [comment]
> In addition, it would be happy if we can simplify `FIFO#read_lines` to make easy to read although I have no idea about it yet. What do you think about these my comments? @daipom

Thanks for your comments!
About debug messages, I have added a fix: 955cee5681f2c1cb375364a421ac1bdc2364e811.

About the simplification, I have the same opinion. I'll think about it for a bit.
If I come up with a good idea, I will add a fix.
Thanks!

**Pull request merged**